packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -167,6 +184,26 @@ export default createRule<Options, MessageId>({
         if (invalidAncestor.type === AST_NODE_TYPES.ReturnStatement) {
           // handle return statement
 
+          if (options.ignoreVoidInVoid) {
+            if (
+              invalidAncestor.parent.parent?.type ===
+                AST_NODE_TYPES.FunctionDeclaration &&
+              invalidAncestor.parent.parent.returnType?.typeAnnotation.type ===
+                AST_NODE_TYPES.TSVoidKeyword
+            ) {
+              return;
+            }
+
+            if (
+              invalidAncestor.parent.parent?.type ===
+                AST_NODE_TYPES.ArrowFunctionExpression &&
```

Comment:
[bug] We should also check function expressions, like:

```ts
declare function returnsVoid(): void

const foo = function (): void {
  return returnsVoid();
}

class Foo {
  foo(): void {
    return returnsVoid();
  }
}

const data = {
  foo(): void {
    return returnsVoid();
  }
}
```

(all the cases above are false positive)

([playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXzg2RlQGcA1HLYACgEoAueANyuACh2w9SMkcc8ALxI0mXPnpNW1eAG928AiCIllqspWr0A3OwC%2BnSFFKl4AMQHzF-HFJZtrSpYWL5XJCm103Dhrjx8wFAYUMJOtvYywBEuKm7qbl7adHpKfkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANrhsORNGgdokADSatWLdkiVhnRQDVJUgJJN701fjiIT2AL7eAupp%2BPkA&tsconfig=&tokens=false))

Reply:
it seems like a case I hadn’t thought of. Thank you.